igvc_perception/src/slam/slam.cpp

```diff
@@ -0,0 +1,396 @@
+
+  g_last_time = msg.header.stamp;
+#if defined(_DEBUG)
+  ROS_INFO_STREAM("Wheel odom reading: " << curr_wheelOdom_reading_.x() << ", " << curr_wheelOdom_reading_.y() << ", "
+                                         << curr_wheelOdom_reading_.theta());
+#endif
+}
+
+/**
+ * Adds the wheelOdom measurement to the factor graph
+ */
+void Slam::addWheelOdomFactor()
+{
+  gtsam::Pose3 betweenPose(gtsam::Rot3::Rz(g_theta), gtsam::Point3(g_x, g_y, 0.0));
+
+  auto factor = gtsam::BetweenFactor<gtsam::Pose3>(
+      X(curr_index_), X(curr_index_ + 1), betweenPose,
+      noiseDiagonal::Sigmas(
+          (gtsam::Vector(6) << g_thetaVariance * 2, g_thetaVariance * 2, g_thetaVariance, g_xVar, g_yVar, g_zVar)
+              .finished()));
+
+  graph_.add(factor);
+
+  g_x = 0;
+  g_y = 0;
+  g_theta = 0;
+  g_xVar = 0;
+  g_yVar = 0;
+  g_zVar = 0;
+  g_thetaVariance = 0;
+}
+
+/**
+ * If there are IMU measurements in the accumulator, this adds them as a single factor to the factor graph and optimizes
+ * graph.
+ */
+void Slam::integrateAndAddIMUFactor()
+{
+  if (accum_.preintMeasCov().trace() != 0)
+  {
+    // Add bias factor
+    auto factor = boost::make_shared<gtsam::BetweenFactor<gtsam::imuBias::ConstantBias> >(
+        B(curr_index_), B(curr_index_ + 1), gtsam::imuBias::ConstantBias(), bias_noise_);
+    graph_.add(factor);
+    graphValues.insert(B(curr_index_ + 1), gtsam::imuBias::ConstantBias());
+
+    // Add imu factor
+    gtsam::ImuFactor imufac(X(curr_index_), V(curr_index_), X(curr_index_ + 1), V(curr_index_ + 1), B(curr_index_ + 1),
+                            accum_);
+
+    graph_.add(imufac);
+    auto newPoseEstimate = result_.at<gtsam::Pose3>(X(curr_index_));  // gtsam::Pose3 newPoseEstimate
+    newPoseEstimate =
+        gtsam::Pose3(accum_.deltaRij() * newPoseEstimate.rotation(), accum_.deltaPij() + newPoseEstimate.translation());
+    graphValues.insert(X(curr_index_ + 1), newPoseEstimate);
+
+    Vec3 last_vel = result_.at<Vec3>(V(curr_index_));
+    last_vel += accum_.deltaVij();
+    graphValues.insert(V(curr_index_ + 1), last_vel);
+
+    curr_index_++;
+    optimize();
+  }
+  accum_.resetIntegration();
+}
+
+/**
+ * Triggers ISAM2 to optimize the current graph and publish the current estimated pose
+ */
+void Slam::optimize()
+{
+#if defined(_DEBUG)
+  static int iteration = 0;
+  ROS_INFO_STREAM("SLAM: Iteration:" << iteration++ << " Imu_updated: " << imu_update_available_
+                                     << " curr_index: " << curr_index_);
+#endif
+  // Adds new factors, updating the ISAM solution and relinearizing as needed.
+  isam_.update(graph_, graphValues);
+
+#if defined(_DEBUG)
+  // Add initial estimates to history_
+  history_.insert(graphValues);
+  graph_.printErrors(history_);
+
+  ROS_INFO_STREAM("printErrors:");
+  isam_.getFactorsUnsafe().printErrors(history_);
+
+  ROS_INFO_STREAM("isam_ graph");
+  isam_.getFactorsUnsafe().print();
+#endif
+
+  result_ = isam_.calculateEstimate();
```

Comment:
Not sure if I should be using [calculateBestEstimate()](https://gtsam.org/doxygen/4.0.0/a03643.html#a9d6f2b0d018f817f64fee6abdaa413ff) instead